### DISCUSSION TITLE
🔎 In the wild! Where are you using XState?

### DISCUSSION THREAD

**Discussion by davidkpiano** [statelyai/xstate#255] [General]
### If you are using XState in a project/at your company, please post it here!

I'm working on a new "In the wild" docs section (related: #172) and I would absolutely ❤️ to highlight projects that are using XState (version 3 or 4).

If you have:
- a (personal?) project in production that is using XState (front-end or back-end)
- OR are at a company that is using XState internally 
- AND you would be happy and willing to share that you're using it

Please post it here. 😄🙏 

Thanks in advance! 💙

**Comment by famulus** [comment]
At https://bridge21.com/ we are re-writing our site using XState 3 and react-automata. It's a game changer. 

Our onboarding is complex with many loops in the process.  The stateMachine pattern reduces fragility, making it way easier to add features without breaking past work. I love being able to see the entire app in the visualizer. 

Our stateMachine has about 90 states so far!  All of them quiet necessary. 

react-automata makes building state dependent interfaces easy `<State is={['*.onboarding.*', '*.returnCustomer.*']}> ... </State>`

**Comment by paul-f-maxson** [comment]
It's not much but I made [a Pomodoro Clock emulator](https://github.com/paul-f-maxson/pomodoro-clock) using React and XState with no other intermediary library. I'm using Hooks to subscribe individual components to an XState service.

**Comment by adamkl** [comment]
Sorry I can't say anything "official" one way or the other, but I thought you'd like to know that we're starting up an internal proof-of-concept at my company to explore the use of XState. 

We're really interested in the potential of pairing up designers and developers (leveraging something like [Sketch](https://sketch.systems/)) to work together on application flows and behaviours. We can see a *lot* of use-cases for this approach. 

Depending on how things turn out, I hope to circle back with more details in the future.

**Comment by davidkpiano** [comment]
@adamkl That's awesome, please keep me in the loop! One of the major goals for XState is to be integrated into design environments, similar to https://overflow.io.

**Comment by davidkpiano** [comment]
Bumping this 😄

**Comment by dhmacs** [comment]
We'll release soon a Shopify app, and we started experimenting with Xstate.. I'll let you know as soon as we ship it and I have an url

**Reply by a-y-u-s-h** [reply]
Any update on this? Is the code going to be open sourced?

**Reply by a-y-u-s-h** [reply]
Nevermind, saw your comment below.

**Comment by jonlaing** [comment]
Working on a POS system with my clients currently. I'll ask them if they'd be cool with me putting it up as an "in the wild" example. I'm sure they'll be fine with it, but I don't want to jump the gun if they have objections.

**Comment by darrenjennings** [comment]
Deployed to production in Kong to handle form states (error, loading, modal, fetch machines etc.) in our Vue app: https://konghq.com/products/kong-manager/

**Comment by hnordt** [comment]
Using XState 4 extensively in [4D Healthware](https://www.4dhealthware.com)'s frontend app.

**Comment by alizbazar** [comment]
We are using XState at [Meru Health](https://www.meruhealth.com) in our React Native app for managing many interactions. I just gave a [presentation of using XState at React Helsinki](https://www.facebook.com/fraktio/videos/780228525664779/?t=4263) - check it out! 😊

**Comment by carloslfu** [comment]
I used **XState** in a touchable app for the **Royal Tyrrell Museum** (Alberta - Canada) and documented its use in this article: ["Modeling a Screensaver with a Statechart, a real use case"](https://medium.com/@carloslfu/modeling-a-screensaver-with-a-statechart-a-real-use-case-f57301682570).

**Comment by bradwoods** [comment]
Used XState for the navigation of a keg analytics web-app [BinaryBeer](https://binarybeer.io/).

**Comment by dhmacs** [comment]
Used XState to model data fetching and filtering of a collection of products 👉 [stimm.com](https://stimm.com/collections/felpe-streetwear)

> Note: Filtering is done client side for platform limitations

**Comment by geddski** [comment]
Using xstate like crazy in https://serviceworkies.com and all my future games.

**Reply by sigfriedCub1990** [reply]
I've seen couple of games you made, they're really good and I'm sure you learned a whole lot by playing them. I'll give `Service Workies` a try since I have some interest in `Service Workers`. Thank you!

**Comment by ZempTime** [comment]
We're now using xstate to power a couple of the more complicated interactions at my place in our Behavior Insights Engine at [Maritz](https://www.maritz.com/).

**Comment by EmirGluhbegovic** [comment]
I'm busy establishing a new company, developing a variety of applications, have removed Redux and Flow from the stack and introduced Typescript and XState. It will add a learning curve for junior devs who join the team, but also teach a rather fundamental lesson instead of just learning a technology (how I feel about Redux).

**Comment by davidkpiano** [comment]
> It will have add a learning curve for junior devs who join the team

That's expected 😅 But I would love to know what the difficult parts are so I can improve documentation and make more examples. Thanks!

**Comment by EmirGluhbegovic** [comment]
> That's expected 😅 But I would love to know what the difficult parts are so I can improve documentation and make more examples. Thanks!

I don't mind putting together different basic scenarios that can be referenced in the documentation to help understand concepts. XState is a powerful library but I feel that it has a lot of possible options without clarity when to implement which possible solution, maybe if you have a solid foundation in computer science and state machines it would be more clear. The objective of the scenarios would be 
to expect users do not have that much experience. That would obviously help the junior devs and anyone else who is learning XState. I am still creating a prototype application to learn XState when done I dont mind sharing that code and breaking it up into different use cases. Let me know what you would suggest etc... 

I would like to see this library grow and be more adopted.

**Comment by davidkpiano** [comment]
One of the things I'm releasing soon is an update to the visualizer (https://statecharts.github.io/xstate-viz) which will allow developers to share visualizations and copy/paste JavaScript/TypeScript machines, and start playing around with them. I feel like playing around with the visualizer is a great way to quickly grasp the concept of state machines and statecharts.

**Comment by davidkpiano** [comment]
@nick-walt let's move this discussion to a new issue: https://github.com/davidkpiano/xstate/issues/552

**Comment by martypenner** [comment]
Using it at [Race Roster](https://raceroster.com) to power a complex (to me) set of client-to-server operations for processing reporting data. The operations have to be sequential, and each state contains data necessary for the next state to operate. The whole thing can be cancelled at any time, or restarted based on a response from the server. An entirely separate state machine can interrupt this one, or this one can kick off the other one. I coordinate both machines with a higher-level machine. Would have taken me **weeks** to walk through all the logic with cancellation without XState. You've made a wonderful tool.

**Comment by linkstrifer** [comment]
Using it in [El Tablero Club Café](https://apps.apple.com/app/el-tablero-club-cafe/id1481988738?src=AppAgg.com) to power some components, i'm migrating every component from useState to xState so is WIP

Want to add model based testing in the future (still trying to understand finite state machines, one step at the time)

**Comment by VinSpee** [comment]
Hi! We're using XState in an increasing number of places on TED.com as well as in some of our conference experience applications.

**Comment by davidkpiano** [comment]
That's awesome @VinSpee! I know that @jonbellah was exploring using it there.

**Comment by VinSpee** [comment]
Yes! It's made it into production a few places now

**Comment by bohendo** [comment]
The state-channel crypto wallet at https://daicard.io (source code is part of the [Indra monorepo](https://github.com/ConnextProject/indra/tree/master/modules/daicard)) is using xstate lightly w plans to utilize it more heavily down the road.

**Comment by andrewgordstewart** [comment]
> The state-channel crypto wallet at https://daicard.io (source code is part of the [Indra monorepo](https://github.com/ConnextProject/indra/tree/master/modules/daicard)) is using xstate lightly w plans to utilize it more heavily down the road.

The "[statechannels project](https://github.com/statechannels/monorepo)" is also currently using xstate heavily to specify how a standard state channel wallet should behave. The daicard wallet, which is based on the counterfactual framework, will be migrating to the standard state channels framework once it's production-ready.

We hope that these specifications can be turned into wallet implementations in different environments!

**Comment by sarimarton** [comment]
Using xstate internally at Morgan Stanley for implementing a mission critical project's auto-save behavior, including debouncing, service error retries and sort of that. More to come, we're buying in. We're interested to see the development of xstate. The visualizer is a must-have and I had to make a custom build with an updated xstate version in it, because a bug in the 4.6.* line affected my machine.

**Comment by davidkpiano** [comment]
@sarimarton We are working on updating the visualizer and add dev tools (which will be much more useful than the visualizer for your workflow) by the end of the year!

**Comment by tmikeschu** [comment]
To elaborate on @jonbellah 's work, my favorite state machine at the moment is used for a digital signage app at TED conferences that transitions slides based on time and server data.

**Comment by 11111000000** [comment]
there is no public link for now, but we use that in booking service prototype

**Comment by viglucci** [comment]
Our team at [Blizzard](https://blizzard.com) leveraged xstate for the first time (via xstate-react) to control the various states on our beta signup panel on the official [World of Warcraft Shadowlands](https://worldofwarcraft.com/en-us/shadowlands/) website.

![image](https://user-images.githubusercontent.com/6305490/78632038-8f548d80-7863-11ea-954d-7e5879987662.png)

**Comment by epolderman** [comment]
> Our team at [Blizzard](https://blizzard.com) leveraged xstate for the first time (via xstate-react) to control the various states on our beta signup panel on the official [World of Warcraft Shadowlands](https://worldofwarcraft.com/en-us/shadowlands/) website.
> 
> ![image](https://user-images.githubusercontent.com/6305490/78632038-8f548d80-7863-11ea-954d-7e5879987662.png)

That's awesome. We have XState in production apps as well, and removed redux.

**Comment by UberMouse** [comment]
I'm using XState at [Koordinates](https://koordinates.com) to power an Electron app we are building. Lots to learn, lots of patterns to figure out, but I'm finding XState great to work with.

First React project I have done without Redux. Just XState + RxJS.

**Comment by mgostisha** [comment]
We've recently started using it at [Flipgrid](https://info.flipgrid.com) across a couple of our repos. Love it ❤️

**Comment by tivac** [comment]
We shipped a game where the UI is almost completely powered by XState, it also drives our svelte component tree using the precursor library to [`xstate-component-tree`](https://github.com/tivac/xstate-component-tree/).

https://playcrucible.com

**Comment by RafalFilipek** [comment]
https://orange.pl 

Right now just in few bigger components like multi step "stuff" browser.
Soon, we will migrate our products comparison page orange.pl/porownywarka

And the biggest (in progress) project - "Sufler"  our online selling widget / system.

Sneak peek :) This is early draft of  top level machine. Each step  (machine) is, more or less quite the same size. Our top level is just responsible for controlling flows between steps.

![image](https://user-images.githubusercontent.com/80960/83688956-83c8dd00-a5ee-11ea-8390-49d7e50ce42a.png)

**Reply by WangLarry** [reply]
wa!!

**Comment by Blacktiger** [comment]
At [https://www.sonatype.com/](Sonatype) we've started using XState for state management as we transition the [repository manager](https://github.com/sonatype/nexus-public/) frontend to React.

**Comment by youmustfight** [comment]
[Upsolve.org](https://upsolve.org) has incorporated **xstate** + **@xstate/test** into our document gathering flows and we're super excited to incorporate it into the rest of our services. We run a web app/service that lets low income Americans file chapter 7 on their own for free (since they can't afford $1k+ attorney fees) and wipe away predatory loans, medical bills, and other debts. XState is definitely going to become the backbone of our complex questionnaires that gather info from folks to fill out their 80+ page paper bankruptcy forms. So appreciative for your work and the all the XState collaborators 💙  

![image](https://user-images.githubusercontent.com/4956240/85149821-bfcf8500-b21f-11ea-8578-285ab523e5c0.png)

**Reply by davidkpiano** [reply]
That's awesome! Love to see `@xstate/test` being used in the wild. Would love to hear about any potential painpoints with questionnaires - a lot of it can be solved easily in userland but I'm thinking about providing some sort of `@xstate/patterns` library to make it even easier.

Cool logo, by the way!

**Reply by youmustfight** [reply]
Not a `@xstate` pain, but to stir up some thoughts the biggest challenge we're facing is around our questionnaire that's hundreds of questions long. I want to give the team a way to quickly skim/read the content of it all so we can easily adjust refine and questions. Right now our non-@xstate implementation requires a team member to click through every screen to view the questions. This makes it really time consuming for folks to review questions at the end. They can't go from birds eye view and then jump into specific parts of the questionnaire. I love the visualizer for showing component interactivity, but we're going to need to figure out another way to render a preview of all the questions/sections that feels more like skimming a text book.

Still thinking through how to do this, and I guess one pain point I hit as a newbie to state machines was knowing when to be nesting machines within one another. In on our document upload flow I avoided it because it made things a bit more complicated (can't remember the snag I hit), but I'm wondering if I should really push for hierarchical state machines when doing that complex flow? With a questionnaire that has sections, is your first instinct that nested machines are ideal or are there some drawbacks/considerations where you think it's better to split them apart?

**Reply by davidkpiano** [reply]
Invoking/spawning actors might be a better fit. Especially if each question in the questionnaire has its own non-trivial logic, it's better to not make the machine too large/complicated.

However, if the questions are simple, and all have the same logic, having the questions as data in `context` makes more sense.

**Reply by youmustfight** [reply]
It has been awhile but our core product has been re-written in XState and its so wonderful 🥳 we're doing spawned actors (thank you for the suggestion!), but treating each question as its own state since the conditionals/content complexity can vary widely. I mentioned this in the docs wish list, but would love to contribute to the patterns docs with some boiled down version of what we're doing if its helpful for folks. I think others in the legal/social services product building space would benefit

**Comment by AlCalzone** [comment]
About 2 1/2 years ago, I started writing a [driver for Z-Wave](https://github.com/AlCalzone/node-zwave-js) controllers using only TypeScript with no native dependencies except `node-serialport`. 
I'm currently in the process of rewriting the core logic to `xstate`, so your baby will soon be powering a bunch of smart homes.
[![logo](https://user-images.githubusercontent.com/17641229/87046180-1ade1180-c1f9-11ea-9dd1-d37afcac8811.png)](https://github.com/AlCalzone/node-zwave-js)

**Reply by davidkpiano** [reply]
That's super exciting! By the way, XState will always have zero dependencies - that's a firm design goal.

**Reply by AlCalzone** [reply]
Good to know!

I've collected a few dependencies since I started, but I try to limit it to the minimum (logging, error reporting, fs-extra, serialport). One goal was to avoid native C++ dependencies as much as possible, since they are often fragile for the end user.

**Comment by shuvalov-mdb** [comment]
Hi!

I developed a C++ code generator for Xstate: 
[https://github.com/shuvalov-mdb/xstate-cpp-generator](https://github.com/shuvalov-mdb/xstate-cpp-generator)

In short, you can model and visualize your State Machine in Xstate, then import my library, call `generateCpp()` method and voila! you'll have fully functional C++ code in your directory.

I hope you can try it and also make people aware of it, I'll be happy to have in production.

-Andrew

**Reply by Andarist** [reply]
Wow! the idea seems awesome ❤️ 

Would you be up for a call to talk about this?

**Reply by carloslfu** [reply]
Very nice! I can't stop thinking about WASM compilation 🤔

**Reply by shuvalov-mdb** [reply]
Yes, we can discuss this on Skype if you need more information. I'm also getting notifications from this thread on my work email, so please discuss here. Not sure how to exchange emails privately, don't want to get spammed...

**Reply by Andarist** [reply]
@shuvalov-mdb feel free to write to me on my email (it's public) or on Twitter (my DMs are open). You can find both on my profile here: https://github.com/Andarist

**Comment by carloslfu** [comment]
We are using XState for coordinating very complex flows on a video-call cross-platform application at @withyourtribe. https://www.withyourtribe.com.

**Comment by ratacat** [comment]
I've been using xstate for some of the components of a bot I'm writing that lives on the ethereum network. I love it! It's been a total game changer for me, and is extremely reliable. I think that as it's closer to completion, it will be substantially more advanced than other similar bots on the network are. Also it's behaviors will be easier to extend and improve upon I think. Thankyou so much @davidkpiano !!

**Comment by JayNaire** [comment]
I retired from electronics & s/w engineering last century.
I am trialling Xstate for its visualisation capabilities and the remorseless burden it puts on me upfront to be very clear in my mind what I want my DIY embedded boiler control system to do at every stage without having to write too much code.
(I used Node Red for prototyping a lot of Home Automation stuff. It's terrific but suffers when developing / debugging systems with many control and failure modes, events and states.)
I hope that Xstate will help me particularly with:
1. Replaying events that drive my system into "weird" modes.
2. Getting the system hardened against brownouts, sensor failures etc.
3. Answering questions like "why are you doing what you're doing right now?" at a glance.
4. Making it easier to change control algorithms if necessary.

Using FSAs in the '80s these kind aspirations were seriously hard work; costlier than the projects I worked on would bear.
With Xstate they look eminently achievable - and fun too!

Thanks and good luck @davidkpiano

**Reply by tomByrer** [reply]
Ironically enough, if I ever teach XState, it will be using 'hardware' examples.

**Comment by moh12594** [comment]
Hello !

At [https://www.frichti.co/](https://www.frichti.co) we also use xstate in order to handle all our different states !
A really good move from Redux for a better readability, a fully declarative approach and a really cleanest code !

Thanks for the big work @davidkpiano and all contributors 👏

![xarTPQAv_400x400](https://user-images.githubusercontent.com/22661810/118690466-60e80680-b808-11eb-95a8-57c817efceb3.jpeg)

**Comment by mrsln** [comment]
At Fugo.ai, we are building a modern Digital Signage CMS. We are using XState in our players ([web version](https://player.fugo.ai)) and in our Design Studio.

**Comment by DanCouper** [comment]
At ThingCo, I've been using it for the past year and a half to manage authentication flows on mobile and web applications (React and React Native). Built a library to wrap the functionality. Sorry for this brain dump but:

It's quite a complex machine, one that has to be configurable per client (OTP or username and password, PIN or biometrics, with various other options that relate to authentication flow/device security). The flows are a sequence of async calls, and I've gone back and forth over the best ways to handle this. Also using Typescript, which I've found to be a double-edged sword w/r/t to XState configuration.

The library has been through four major iterations now, and I'm tended to agree with most of the observations made in [this critical analysis of statecharts (PDF link)](https://mbreen.com/breenStatecharts.pdf). I think XState is the gold standard for state machines/statecharts in JS/TS, and I can't see myself using anything else in similar situations to the complex sequential flows I've built, but there's been quite a bit of pain.

So the library is structured with a core machine that controls the auth state. That is created via a factory function that allows me to inject config at build time (or in test apps, from a parent React context that does the same thing). That machine is interpreted in a context provider, accessed via hooks in-app. States essentially represent "screens" in the apps.

Initially, I wrote a single machine that made use of invoked promises. This rapidly became unmanageable. Context grew, states exploded.

So second, third iterations split out functionality. The second iteration used nested machines. This caused severe TS typing issues unless I inlined everything (which defeats the point somewhat).

The third iteration used actors, which enabled some nice properties. The context had become pretty populous, and I could reduce that to almost nothing, keeping values within actors. The parent machine could just have an overall state, and that would be what the app used for the app screens, and things like loading/error states remained within the actors. So for example, when making the initial sign in, the actor for the username service starts, it has an idle state for awaiting input, then on input, switch to a loading state, then on success store the response object, tell the parent that was complete, etc etc.

This _worked_, but instead of a state explosion, I got an event explosion. Events had to be defined for both parent and child machines. The number of actions also exploded, because I had lots of small contexts to manage + the parent context. Loading was a problem: there are about 20 different possible async calls, and each time one was invoked, the parent had to be notified. When it was resolved, the parent had to be notified. When there was an error, the parent had to be notified. Then I had to manage starting and stopping of actors. Started using enums extensively to define state ids, action names, guard names, event types, which was horrible but prevented mistakes. Responses from invoked promises have their own internal event types, so all those had to be typed. It all _worked_ and worked well, but it was a nightmare from a developer PoV.

About this point, the `model` API came in, which has been really useful from a TS point of view. _However_, there are three main issues I've had with it:
- the state IDs are no longer defined up-front and so I get no type inference. With complex machines, this is a major annoyance. Due to this, state IDs are the only thing I've retained enums in the codebase for.
- the event creators I liked at first but dropped them in favour of just defining a union type and passing it into the model's generic parameters. The issue is that the type often (depending on where in the codebase I'm using it) isn't just a simple `{ type: "MAI_EVENT" }`, it's a nasty `FinalEventCreators<>` thing that's difficult to work with. Just using plain objects ended up being far easier.
- was still getting loads of type issues, so reverted to using `createMachine<typeof Model` -- I realise this option has been removed in the latest release so I'll need to do some tweaking when I upgrade (hopefully most of the issues I've had re typings have been nixed by now).

So final iteration.
- Dumped actors and just have a single state machine. This means no need to manage multiple machines and the message passing required to herd them, much easier to handle from a cognitive perspective.
- Dumped invocations. This means there is no need to carefully handle the responses within the machine. It also fixed a huge issue we had, which is that putting the async calls within the machine meant _a._ that I had to build a facade for the APIs we use and pass that into the machine factory, and _b._ that making the calls within the machine logic effectively hid everything and made it incredibly hard to debug -- things would just silently fail, and it was often almost impossible to figure out why as all the logic was hidden.
- Dumping invocations means no loading states, which is great

So one machine, almost standard-traffic-light-example simple. Then each state has a corresponding hook (because the state IDs are string enums, I can then use string templating in TS to ensure all states are covered and all states have corresponding selector functions). The hooks accept a callback, which is the API call relevant to that state. This makes debugging them very easy, allows for mock test versions to be injected, and dispenses with having to make API facades. The hooks all return a primary method, which is a function that takes the callback passed to the hook fires it, sending the correct events into the machine at the correct times. This also moves loading/validation errors etc out of the machine context.

Overall, model testing is very good if a bit slow -- @xstate/test works excellently, I've got it set up with Jest/React testing library to thoroughly exercise the machine and the hooks that trigger all the machine events, and it's great. 

The new visualiser stuff is very nice (although running in watch mode when developing and it switching the tab to the visualisation on every code change has been driving me slowly insane for the past three months).

Library is great, anyway, apologies for the wall of text. Work I've done is internal, but I may be able to share the structure we have as it doesn't contain any business-specific  logic (it's mainly all AWS Cognito-related stuff), I'll have  a check

**Reply by ChrisShank** [reply]
Thanks for the detailed, thoughtful feedback! If you have problems with TS or modeling always feel free to reach out on the discord!

**Reply by mattpocock** [reply]
@DanCouper You might get use out of the VSCode extension! This means you don't need to swap tabs, and can also validate state names for you (no more enums!)

https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode

**Reply by mattpocock** [reply]
Also, would this assist with the event explosion? It means that updates down the tree between spawned machines can be more declarative:

https://twitter.com/mpocock1/status/1440356513332162575

**Reply by DanCouper** [reply]
@mattpocock sorry for late reply, but the VSCode extension looks good so far. Wee bit of an issue that I've just switched to vim as my primary driver but hey ho. I think my main issue (bar annoyance) with the inspector's current behaviour is that from a DX PoV I'd like to simply turn it on or off, which turns out to be quite difficult (and I don't *really* want to spend time writing a version of the inspector just for my project). This should be useful as a sticking plaster for that though!

Re that `useQuery` proposal, that does look really interesting, it may well salve some of my misgivings. I want to revisit spawning machines, dig into it a bit further, but it's still clunky atm and I'm not sure the extra complexity is worth it. I use Erlang (well, Elixir) a bit, which coincidentally uses something similar to the actor model (as mentioned in the XState docs). But with that, there's a toolbox of stuff to handle processes if needed (and generally explicit FSMs aren't needed due to the pattern matching capabilities of the language). For example, everything is set up and supervised in the same way, there is a Registry module that acts as a process ID library. Whereas XState doesn't have these tools or conventions; I mean, it's just a library not an ecosystem, but at the same time I feel like it needs some tools in place to help if an actor-like model is being used -- `useQuery`, though obviously React-specific looks like one of those things

**Reply by zacharyweidenbach** [reply]
I appreciate this post, as I'm on the lookout for cautionary tales with xstate, because it seems like a good way of standardizing UI architecture on a team, though it can feel "boiler plate-y" for simple UIs.

In your case, I wonder if the primary source of your woes was that network response caching responsibilities were entangled with application and business logic. From what I could understand, you have a machine that controls the overall state that your UI should be in. But on top of that, you added the responsibility of managing api response data and errors (by storing it in the context of the machine). I wonder if you delegated that responsibility to something else, you could have avoided the spiraling complexity that you encountered. In a way, it sounds like that's ultimately what you did with the hooks. The hooks took on the responsibility of both choreographing the network requests, and sending an event back to the machine to move on. Conceptually this feels odd, because now your business logic for the network requests live in your view layer (react) and the machine is simple to the point that it could be replaced with react's `useReducer` instead (although you don't get model based testing or state charts).

I hit a similar stumbling block when trying to build a network layer with xstate. My problem was, if a machine's primary responsibility is to manage the UI of a screen, and its lifecycle is tied to that of the UI, it needs to access an api cache that predates it, and lives on after it stops (when the user leaves that screen, the component unmounts, and the machine stops). When I tried to delegate network fetch and caching responsibilities to another child machine, the parent machine invokes or spawns this child network machine, but the fetch response/error data maintained in the context of this child machine lives and dies with this child machine, as you probably found. I could send the api response data or error upward to the parent machine to manage itself. But if I have many requests happening on a given screen, then that becomes unwieldy (as you found). Worse still, if I have api response data that needs to be displayed on my UI in a more global component (like a user's first/last name in the nav bar) and it is edited by a fetch request in a deeply nested machine, I now have a really difficult problem of propagating that data out to another machine higher in the component tree. 

Instead, if I delegated the storing of api responses and errors to something else that wasn't tied to my UI machine's lifecycle I could strike the right balance, where the machine controlling my UI's state could invoke the fetch promises as services (through a constrained interface exposed to it from my network layer). But the actual storing of the response or error lived in the network layer, and could be accessed directly in the UI machine via the interface exposed by the network layer. On the react side, the component could subscribe to the same api data/error via a hook (also exposed by the network layer), so that it could rerender when the data changed, just like you would if it were coming from context or redux or any other reactive source. The network layer functions as the single source of truth and provides a means for machines and react components to access the api response cache.

In your situation, you'd still need to configure what to do if an error occurs, but the `onError` as a result of the invoke service feels like a pretty natural way to deal with errors, now that the overhead of storing those errors is managed elsewhere.

Apologies ahead of time if I misdiagnosed the true root of the problems you faced.

**Reply by dvanderweele** [reply]
Realize I'm necroposting, but there are a few VIM extensions for VS Code that work okay. Nothing beats the real neovim of course, but even so.

**Comment by GuillaumeAmat** [comment]
We also use XState at [Back Market](https://backmarket.com) 💚

The library itself, that lovely editor of yours and probably soon, the visualizer embedded in back-office pages, to better share the implemented logic with the stakeholders.

Looking forward to the awesome v5 in the making :raised_hands:

**Comment by zacharyweidenbach** [comment]
I've recently refactored the bootstrap and authentication phases of a react-native app with great success. I've worked on this app on and off for the last 3 years, and this is the first time this area of the app feels organized and manageable.

To give more background, the bootstrap phase of the app has some unfortunate complexity due to design decisions made by the backend team. The api requires a session cookie in order to recover cached details about an authenticated user. The unfortunate thing about session cookies between iOS and Android, is that iOS will forget them after the app is force closed or garbage collected, whereas Android won't. This is because of the underlying implementation of session cookies between platforms. iOS uses Safari's implementation while Android uses Chrome's. So, the problem on iOS was that the app would open fresh with an access token in it's cold store, and try to use this token to make authenticated requests, but would not have a session cookie. The api would see the request as authenticated, but assign a new session cookie to the client, and subsequent requests to other authenticated endpoints would fail because the api could not recover the necessary cached data on this newly assigned session cookie that wasn't properly associated with an authenticated session. Phew.

So now the app needed to open, check if it had a session cookie in its cold storage, if it did it would rehydrate this into the device's cookies.

Next, the app needed to make a call to a configuration endpoint to get instructions from the api for how to authenticate (which IDP host to call, what client id to use, etc). If the app didn't have a session cookie, this is when it would be assigned one. After this call, the app needed to persist this cookie into cold storage for safe keeping. On top of this, there were various other setup events that needed to occur, such as configuring a background geolocation plugin, etc.

With xstate, this sequence of events could be very explicitly modeled as a state machine that does each async step one after the other in a `BootstrapMachine`. It now reads like a recipe of stepwise events that are all colocated, and a failure at any point along the way can have very clear error logging, and failing forward to the next step. 

We have also just introduced codepush, and we need to polish the user experience to show them some UI feedback while codepush installs a new OTA update and restarts the app. It is super easy to visualize how this can live in the BootstrapMachine, where the app can be made to show this codepush OTA download/install UI feedback before restarting, checking that it has the most up-to-date codepush version, and continuing on to mount the rest of the normal app.

Then there's the authentication sequence in the app, which requires 3 different endpoint calls before we can consider the user authenticated. Previously, this was managed by a series of react component containers that would conditionally mount their children after a given request was completed. This led to the request logic being spread across several components, and made working on anything to do with authentication a painful experience. With xstate, the complexity of this sequence of calls was drastically simplified. The retry logic, and logout on failure of any of the necessary requests because super simple to visualize, handle, extend, add additional analytics, you name it!

XState really did wrestle the complexity of these various stages of the app into a very manageable, extendable, readable list of steps. I really can't emphasize how long I've struggled to find a better solution to this complexity. I started with redux and sagas, then `react-query`, and finally this. It finally feels manageable.

**Comment by bmadone** [comment]
I use this in my pet-project (aka clone of Toggl, Clockify) for Timer feature, code is open sourced and can be found here https://github.com/bmarvinb/moggl/blob/main/src/features/timer/machines/TimerMachine.ts

And this is how machine looks like
![telegram-cloud-photo-size-2-5382034479830386161-y](https://user-images.githubusercontent.com/23120972/197534879-8a2588ad-1517-4881-a703-2106ad8c8b80.jpg)

**Comment by hsndmr** [comment]
https://github.com/hsndmr/xstate-quiz-app
This is a quiz app built with xstate and react. It is a simple quiz app that has 10 questions and the user can choose one of the 4 options for each question. The user can also go back to the previous question and change their answer. End of the quiz, the user can see their score and can restart the quiz.

**Comment by UberMouse** [comment]
The newly launched [koordinates v10 UI](https://koordinates.com/data/) is almost entirely built with xstate and [xstate-tree](https://github.com/koordinates/xstate-tree/)

**Comment by hsndmr** [comment]
https://github.com/hsndmr/nestjs-auth-starter
The NestJS Auth Starter project is a starter project for [NestJS](https://nestjs.com/) that provides authentication and authorization functionality out of the box. It is built on top of the popular NestJS framework and uses the [mongoose](https://mongoosejs.com/) library to access the database. For authentication and authorization, it relies on the powerful [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken) library, while password hashing is performed using [bcrypt](https://www.npmjs.com/package/bcrypt). Additionally, the project utilizes the [xstate](https://xstate.js.org/) library for state management in authentication cases, and includes tests for all cases. With this starter project, developers can quickly set up a secure and robust authentication system in their NestJS applications.

**Comment by radiodario** [comment]
Hey, i thought i'd contribute to this - we are using xstate on [Anyone](callinganyone.com) (closed source) - a voice networking app with 5 minute calls written in react native. We use it for the whole phone call flow, signup process, matchmaking for in-app events, post-call reviewing of users, etc. We're so grateful for all the hard work on this awesome library - and the efforts of the community!

some of the machines we have:

<img width="1527" alt="Screenshot 2023-03-14 at 23 27 31" src="https://user-images.githubusercontent.com/780409/227995384-9af4cb46-3a55-47e1-922c-1a70fbe0d880.png">
<img width="706" alt="Screenshot 2023-03-14 at 23 27 02" src="https://user-images.githubusercontent.com/780409/227995411-725d8aba-ed1b-43e3-97ab-6c6eb7fe5307.png">
<img width="1562" alt="Screenshot 2023-03-14 at 23 25 55" src="https://user-images.githubusercontent.com/780409/227995458-e44d13d6-f366-4130-940a-22936de6206e.png">
<img width="1590" alt="Screenshot 2023-03-14 at 23 23 35" src="https://user-images.githubusercontent.com/780409/227995486-2446ebdb-aeb2-4d63-bb5d-7ca75861ec22.png">

**Reply by davidkpiano** [reply]
This is awesome, thanks for sharing!

**Comment by gavination** [comment]
These projects are all brilliant to see! Bumping this thread to breathe some life back into it!

**Comment by oscarmarina** [comment]
Hi !

The **XstateReactiveController** is a Lit Reactive Controller specifically designed for straightforward integration with XState.

- https://github.com/oscarmarina/XstateController

- https://stackblitz.com/github/oscarmarina/XstateController?file=README.md

**Reply by christophe-g** [reply]
@oscarmarina  - thanks a lot for posting this - was just about to create the same reactive controller !

**Reply by oscarmarina** [reply]
Hi @christophe-g, I'm pleased to hear that the code is helpful for you! 🚀 I wanted to let you know that I've just uploaded a new version. Now, the Lit controller is consumed as an independent package on NPM. Feel free to check it out and let me know if you have any feedback! (stable version ^5.2.1)

- https://github.com/oscarmarina/XstateController

- https://github.com/oscarmarina/blockquote-web-components/tree/main/packages/controllers/blockquote-controller-xstate
